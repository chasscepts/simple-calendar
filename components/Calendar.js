import { StyleSheet, Text, View } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';

const COLOR = '#666';

const CELL_WIDTH = `${100 / 7}%`;

const styles = StyleSheet.create({
  container: {

  },
  title: {
    fontWeight: 'bold',
    fontSize: 28,
    paddingTop: 5,
    paddingStart: 10,
    color: COLOR,
  },
  headerRow: {
    flexDirection: 'row',
    flexWrap: 'nowrap',
    overFlowX: 'hidden',
    paddingTop: 10,
    paddingBottom: 5,
    width: '100%',
  },
  header: {
    textAlign: 'center',
    fontWeight: 'bold',
    fontSize: 20,
    color: COLOR,
    width: CELL_WIDTH,
  },
  datesPanel: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    width: '100%',
  },
  cell: {
    width: CELL_WIDTH,
    height: 75,
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  emoji: {
    fontSize: 32,
  },
  date: {
    color: COLOR,
    fontSize: 20,
  }
});

const weekdays = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

const months = [
  'January', 'February', 'March', 'April',
  'May', 'June', 'July', 'August',
  'September', 'October', 'November', 'December',
];

const emojis = {
  happy: '🙂',
  fair: '😐',
  sad: '😡',
};

const getMonthDates = (month, year) => {
  const firstDayWeekDay = new Date(year, month, 1).getDay();
  // 0th day of next month is last day of this month.
  const lastDayOfMonth = new Date(year, month + 1, 0);
  const daysInMonth = lastDayOfMonth.getDate();

  const dates = [];

  let startDate = 1;
  let emptyCellsAtStart = 0;

  if (firstDayWeekDay === 0) {
    emptyCellsAtStart = 6;
    startDate = 2;
  } else if (firstDayWeekDay > 1) {
    emptyCellsAtStart = firstDayWeekDay - 1;
  }

  if (emptyCellsAtStart > 0) {
    for (let i = 0; i < emptyCellsAtStart; i += 1) {
      dates.push(0);
    }
  }

  for (let i = startDate; i <= daysInMonth; i += 1) {
    dates.push(i);
  }

  return dates;
};

const Cell = ({ date, mood, isToday }) => {
  if (date === 0) {
    return <View style={styles.cell} />
  }

  return (
    <View style={styles.cell}>
      {mood && (
        <Text accessibilityRole="image" accessibilityLabel={mood} style={styles.emoji}>
          {emojis[mood]}
        </Text>
      )}
      <Text style={[styles.date, { fontWeight: isToday? 'bold' : 'normal' }]}>
        {date}
      </Text>
    </View>
  );
};

const Calendar = ({ dates, month, year }) => {
  //  Assumption: month props is 1 indexed.
  const monthIndex = month - 1;

  const moods = dates.reduce((prev, current) => {
    prev[current.date] = current.mood;
    return prev;
  }, Object.create(null));

  const today = new Date();
  const isSameMonth = today.getMonth() === monthIndex && today.getFullYear() === year;
  const date = today.getDate();

  return (
    <SafeAreaView>
      <View>
        <Text style={styles.title}>
          {`${months[monthIndex]}, ${year}`}
        </Text>
      </View>
      <View style={styles.headerRow}>
        {weekdays.map((day) => <Text style={styles.header}>{day}</Text>)}
      </View>
      <View style={styles.datesPanel}>
        {getMonthDates(monthIndex, year).map((d) => (
          <Cell
            date={d}
            mood={moods[d]}
            isToday={isSameMonth && (d === date)}
          />
        ))}
      </View>
    </SafeAreaView>
  );
};

export default Calendar;
