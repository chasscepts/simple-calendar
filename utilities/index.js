export const weekdays = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

export const months = [
  'January', 'February', 'March', 'April',
  'May', 'June', 'July', 'August',
  'September', 'October', 'November', 'December',
];

export const emojis = {
  happy: '🙂',
  fair: '😐',
  sad: '😡',
};

/**
 * Fills an array with dates of the month.
 * This array behaves like a flattened 2-d array with Sunday as the last column.
 * Empty starting cells (if any) are filled with zeros.
 * @param {number} month 0 indexed month
 * @param {number} year year under consideration
 */
export const getMonthDates = (month, year) => {
  const firstDayWeekDay = new Date(year, month, 1).getDay();
  // 0th day of next month is last day of this month.
  const lastDayOfMonth = new Date(year, month + 1, 0);
  const daysInMonth = lastDayOfMonth.getDate();

  const dates = [];

  let startDate = 1;
  let emptyCellsAtStart = 0;

  if (firstDayWeekDay === 0) {
    emptyCellsAtStart = 6;
  } else if (firstDayWeekDay > 1) {
    emptyCellsAtStart = firstDayWeekDay - 1;
  }

  if (emptyCellsAtStart > 0) {
    for (let i = 0; i < emptyCellsAtStart; i += 1) {
      dates.push(0);
    }
  }

  for (let i = 1; i <= daysInMonth; i += 1) {
    dates.push(i);
  }

  return dates;
};
