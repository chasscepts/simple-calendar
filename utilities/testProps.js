export const dates = [
  {
    date: 4,
    mood: 'sad',
  },
  {
    date: 1,
    mood: 'sad',
  },
  {
    date: 4,
    mood: 'happy',
  },
  {
    date: 6,
    mood: 'fair',
  },
  {
    date: 27,
    mood: 'sad',
  }
];

export const month = 9;

export const year = 2022;
