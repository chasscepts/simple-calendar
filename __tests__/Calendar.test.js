import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import Calendar from '../components/Calendar';
import { weekdays } from '../utilities';
import { dates, month, year } from '../utilities/testProps';
import { View } from 'react-native';

describe('Calendar', () => {
  describe('snapshot', () => {
    beforeAll(() => {
        jest.useFakeTimers('modern');
        jest.setSystemTime(new Date(2022, 8, 14));
    });

    afterAll(() => {
        jest.useRealTimers();
    });

    it('renders correctly', () => {
      const tree = renderer.create(<Calendar dates={dates} month={month} year={year} />);
      expect(tree.toJSON()).toMatchSnapshot();
      tree.unmount();
    });
  });

  it('displays the month and year', () => {
    const testID = 'calendar';
    const { queryByTestId } = render(
      <View testID={testID}>
        <Calendar dates={dates} month={month} year={year} />
      </View>
    );
    expect(queryByTestId(testID)).toHaveTextContent('September, 2022');
  });

  it('displays the weekday headings', () => {
    const testID = 'calendar';
    const { queryByTestId } = render(
      <View testID={testID}>
        <Calendar dates={dates} month={month} year={year} />
      </View>
    );
    const node = queryByTestId(testID);
    weekdays.forEach((w) => (
      expect(node).toHaveTextContent(w)
    ))
  });

  it('displays the 30 days of september', () => {
    const testID = 'calendar';
    const { queryByTestId } = render(
      <View testID={testID}>
        <Calendar dates={dates} month={month} year={year} />
      </View>
    );
    const node = queryByTestId(testID);
    [...Array(30)].map((_, idx) => idx + 1).forEach((num) => (
      expect(node).toHaveTextContent(num)
    ))
  });
});
