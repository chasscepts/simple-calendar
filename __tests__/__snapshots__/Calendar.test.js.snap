// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar snapshot renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 28,
          "fontWeight": "bold",
          "paddingStart": 10,
          "paddingTop": 5,
        }
      }
    >
      September, 2022
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "overFlowX": "hidden",
        "paddingBottom": 5,
        "paddingTop": 10,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
          "width": "14.285714285714286%",
        }
      }
    >
      Mon
    </Text>
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
          "width": "14.285714285714286%",
        }
      }
    >
      Tue
    </Text>
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
          "width": "14.285714285714286%",
        }
      }
    >
      Wed
    </Text>
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
          "width": "14.285714285714286%",
        }
      }
    >
      Thu
    </Text>
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
          "width": "14.285714285714286%",
        }
      }
    >
      Fri
    </Text>
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
          "width": "14.285714285714286%",
        }
      }
    >
      Sat
    </Text>
    <Text
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
          "width": "14.285714285714286%",
        }
      }
    >
      Sun
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        accessibilityLabel="sad"
        accessibilityRole="image"
        style={
          Object {
            "fontSize": 32,
          }
        }
      >
        üò°
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        1
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        2
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        3
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        accessibilityLabel="happy"
        accessibilityRole="image"
        style={
          Object {
            "fontSize": 32,
          }
        }
      >
        üôÇ
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        4
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        5
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        accessibilityLabel="fair"
        accessibilityRole="image"
        style={
          Object {
            "fontSize": 32,
          }
        }
      >
        üòê
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        6
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        7
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        8
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        9
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        10
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        11
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        12
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        13
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "bold",
            },
          ]
        }
      >
        14
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        15
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        16
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        17
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        18
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        19
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        20
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        21
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        22
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        23
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        24
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        25
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        26
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        accessibilityLabel="sad"
        accessibilityRole="image"
        style={
          Object {
            "fontSize": 32,
          }
        }
      >
        üò°
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        27
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        28
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        29
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 75,
          "justifyContent": "flex-end",
          "width": "14.285714285714286%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#666",
              "fontSize": 20,
            },
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        30
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
